src/content/docs/develop/state-management.mdx

```diff
@@ -2,32 +2,180 @@
+
+> Contrary to popular belief, it is ok and often preferred to use the ordinary Mutex from the standard library in asynchronous code ... The primary use case for the async mutex is to provide shared mutable access to IO resources such as a database connection.
+
+It's a good idea to read the linked documentation page fully to understand the trade-offs between the two. However, if you are using `async` commands and want to use Tokio's async `Mutex`, you can set it up the same way and access the state like this:
+
+```rust
+use tauri::{async_runtime::Mutex, State};
+
+#[tauri::command]
+async fn increase_counter(state: State<'_, Mutex<AppState>>) -> Result<u32, ()> {
+  let mut state = state.lock().await;
+  state.counter += 1;
+  Ok(state.counter)
+}
+```
+
+Note that the return type must be `Result` if you use asynchronous commands.
+
+### Outside of Commands
+
+Sometimes you may need to access the state outside of commands, such as in a different thread or in an event handler like `on_window_event`. In such cases, you can use the `state()` method of the `AppHandle` to get the state:
 
 ```rust
-use std::{collections::HashMap, sync::Mutex};
-use tauri::{Manager, plugin::Builder};
+use tauri::{Builder, GlobalWindowEvent, Manager};
 
 #[derive(Default)]
-struct Store(Mutex<HashMap<usize, String>>);
+struct AppState {
+  counter: u32,
+}
 
-Builder::new("<plugin-name>")
-  .setup(|app, api| {
-    app.manage(Store::default());
+// In an event handler:
+fn on_window_event(event: GlobalWindowEvent) {
+    // Get a handle to the app so we can get the global state.
+    let app_handle = event.window().app_handle();
+    let state = app_handle.state::<Mutex<AppState>>();
 
-    // retrieve the store later
-    let store = app.state::<Store>();
+    // Lock the mutex to mutably access the state.
+    let mut state = state.lock().unwrap();
+    state.counter += 1;
+}
+
+// In `fn main()`:
+Builder::default()
+  .setup(|app| {
+    app.manage(Mutex::new(AppState::default()));
     Ok(())
   })
+  .on_window_event(on_window_event)
+  .run(tauri::generate_context!())
+  .unwrap();
+```
+
+This method is useful when you cannot rely on command injection. For example, if you need to move the state into a thread where using an `AppHandle` is easier, or if you are not in a command context.
+
+## Mismatching Types
+
+:::caution
+If you use the wrong type for the `State` parameter, you will get a runtime panic instead of compile time error.
+
+For example, if you use `State<'_, AppState>` instead of `State<'_, Mutex<AppState>>`, there won't be any state managed with that type.
+:::
+
+If you prefer, you can wrap your state with a type alias to prevent this mistake:
+
+```rust
+use std::sync::Mutex;
+
+#[derive(Default)]
+struct AppStateInner {
+  counter: u32,
+}
+
+type AppState = Mutex<AppStateInner>;
 ```
 
-</Stub>
+However, make sure not to wrap it in a `Mutex` twice if you do this, otherwise you will run into the same issue.
```

Comment:
```suggestion
However, make sure to use the type alias as is, and not wrap it in a `Mutex` a second time, otherwise you will run into the same issue.
```